tests/test_updater_delegation_graphs.py

```diff
@@ -0,0 +1,246 @@
+            ],
+            visited_order=["A", "C", "D"],
+        ),
+        "two-level ignores all branches not matching 'paths'": DelegationsTestCase(
+            delegations=[
+                TestDelegation("targets", "A", paths=["*.py"]),
+                TestDelegation("targets", "B"),
+                TestDelegation("A", "C"),
+            ],
+            visited_order=["B"],
+        ),
+        "three-level ignores all branches not matching 'paths'": DelegationsTestCase(
+            delegations=[
+                TestDelegation("targets", "A"),
+                TestDelegation("targets", "B"),
+                TestDelegation("A", "C", paths=["*.py"]),
+                TestDelegation("C", "D"),
+            ],
+            visited_order=["A", "B"],
+        ),
+        "cyclic graph": DelegationsTestCase(
+            delegations=[
+                TestDelegation("targets", "A"),
+                TestDelegation("targets", "B"),
+                TestDelegation("B", "C"),
+                TestDelegation("C", "D"),
+                TestDelegation("D", "B"),
+            ],
+            visited_order=["A", "B", "C", "D"],
+        ),
+        "two roles delegating to a third": DelegationsTestCase(
+            delegations=[
+                TestDelegation("targets", "A"),
+                TestDelegation("targets", "B"),
+                TestDelegation("B", "C"),
+                TestDelegation("A", "C"),
+            ],
+            # Under all same conditions, 'C' is reached through 'A' first"
+            visited_order=["A", "C", "B"],
+        ),
+        "two roles delegating to a third different 'paths'": DelegationsTestCase(
+            delegations=[
+                TestDelegation("targets", "A"),
+                TestDelegation("targets", "B"),
+                TestDelegation("B", "C"),
+                TestDelegation("A", "C", paths=["*.py"]),
+            ],
+            # 'C' is reached through 'B' since 'A' does not delegate a matching pattern"
+            visited_order=["A", "B", "C"],
+        ),
+    }
+
+    @utils.run_sub_tests_with_dataset(graphs)
+    def test_graph_traversal(self, test_data: DelegationsTestCase) -> None:
+        """Test that delegated roles are traversed in the order of appearance
+        in the delegator's metadata, using pre-order depth-first search"""
+
+        try:
+            exp_files = [*TOP_LEVEL_ROLE_NAMES, *test_data.visited_order]
+            exp_calls = [call(role, 1) for role in test_data.visited_order]
+
+            sim = self._init_repo(test_data.delegations)
+            updater = self._init_updater(sim)
+            # Call explicitly refresh to simplify the expected_calls list
+            updater.refresh()
+            # Check that metadata dir contains only top-level roles
+            self._assert_files_exist(TOP_LEVEL_ROLE_NAMES)
+
+            with patch.object(
+                sim, "_fetch_metadata", wraps=sim._fetch_metadata
+            ) as wrapped_fetch:
```

Comment:
Once the fetcher statistics PR is in we should modify this to use them ... but it's not in quite yet